src/iceberg/util/visit_type.h

```diff
@@ -0,0 +1,121 @@
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#pragma once
+
+#include "iceberg/result.h"
+#include "iceberg/type.h"
+#include "iceberg/util/checked_cast.h"
+#include "iceberg/util/unreachable.h"
+#include "iceberg/util/visitor_generate.h"
+
+namespace iceberg {
+
+#define TYPE_VISIT_INLINE(TYPE_CLASS)                                   \
+  case TYPE_CLASS##Type::kTypeId:                                       \
+    return visitor->Visit(                                              \
+        iceberg::internal::checked_cast<const TYPE_CLASS##Type&>(type), \
+        std::forward<ARGS>(args)...);
+
+/// \brief Calls `visitor` with the corresponding concrete type class
+///
+/// \tparam VISITOR Visitor type that implements Visit() for all Iceberg types.
+/// \tparam ARGS Additional arguments, if any, will be passed to the Visit function after
+/// the `type` argument
+/// \return Status
+///
+/// A visitor is a type that implements specialized logic for each Iceberg type.
+/// Example usage:
+///
+/// ```
+/// class ExampleVisitor {
+///   Status Visit(const IntType& type) { ... }
+///   Status Visit(const LongType& type) { ... }
+///   ...
+/// }
+/// ExampleVisitor visitor;
+/// VisitTypeInline(some_type, &visitor);
+/// ```
+
+template <typename VISITOR, typename... ARGS>
+inline Status VisitTypeInline(const Type& type, VISITOR* visitor, ARGS&&... args) {
+  switch (type.type_id()) {
+    ICEBERG_GENERATE_FOR_ALL_TYPES(TYPE_VISIT_INLINE);
+    default:
+      break;
+  }
+  return NotImplemented("Type not implemented");
+}
+
+#undef TYPE_VISIT_INLINE
+
+#define TYPE_VISIT_INLINE(TYPE_CLASS)                          \
+  case TYPE_CLASS##Type::kTypeId:                              \
+    return std::forward<VISITOR>(visitor)(                     \
+        internal::checked_cast<const TYPE_CLASS##Type&>(type), \
+        std::forward<ARGS>(args)...);
+
+/// \brief Call `visitor` with the corresponding concrete type class
+/// \tparam ARGS Additional arguments, if any, will be passed to the Visit function after
+/// the `type` argument
+///
+/// Unlike VisitTypeInline which calls `visitor.Visit`, here `visitor`
+/// itself is called.
+/// `visitor` must support a `const Type&` argument as a fallback,
+/// in addition to concrete type classes.
+///
+/// The intent is for this to be called on a generic lambda
+/// that may internally use `if constexpr` or similar constructs.
+template <typename VISITOR, typename... ARGS>
+inline auto VisitType(const Type& type, VISITOR&& visitor, ARGS&&... args)
+    -> decltype(std::forward<VISITOR>(visitor)(type, args...)) {
+  switch (type.type_id()) {
+    ICEBERG_GENERATE_FOR_ALL_TYPES(TYPE_VISIT_INLINE);
+    default:
+      internal::Unreachable("Type not implemented");
```

Comment:
The docstring says a fallback is called but the body aborts in the default case?

Reply:
TBH I have no idea what that sentence actually means, I don't think these words does any help to understand the template, so I'd rather delete that, or do you have any suggestions?

Reply:
So this was copied from Arrow? It may be worth noting that in the header just for reference.

Looking at it again, I think what it means is that because the visitor here does not inherit from a base class that has all cases defined, you must make sure to handle all cases. So I guess the implementation is OK. But the wording could be improved.

Reply:
> So this was copied from Arrow? It may be worth noting that in the header just for reference.

Yeah, it's adapted from arrow, I'll add the note.

> 
> Looking at it again, I think what it means is that because the visitor here does not inherit from a base class that has all cases defined, you must make sure to handle all cases. So I guess the implementation is OK. But the wording could be improved.

How about we keep this as is for now, and revisit it later, either improve the wording ourselves or sync with Arrow if they update their version.